### PR TITLE
Project Recovery - Catch top level exceptions

### LINKED ISSUE #22901: Project Recovery throws unexpected exceptions
**Original reporter:** Peter Baker/Anthony Lim

### Expected behavior

If project recovery fails to reconstruct the history, it should simply throw an error saying that recovery was not possible.

### Actual behavior

Failure of project recovery results in an unexpected error

### Steps to reproduce the behavior

* Make sure the recovery directory is empty and Mantid is not running
* Unzip the file below to the project recovery directoy
* Start Mantid
* When offered recovery choose yes
* Mantid throws an unexpected error
[2018-07-12T16-33-16.zip](https://github.com/mantidproject/mantid/files/2189268/2018-07-12T16-33-16.zip)

### PR DESCRIPTION
**Pull request by DavidFair**
**Cannot be merged before #22887**

**Description of work.**
Adds a top level try / catch around the recovery attempt. This will open a dialog box informing the user a full recovery could not be completed and that Mantid will continue running correctly.

**To test:**
- Open Mantid
- Create a sample workspace
- Ensure a checkpoint is saved
- Crash Mantid
- On the most recent checkpoint, open a .py file
- Remove the timestamp (everything after the # char)
- Restart and attempt recovery

Partly deals with #22901.

Does this update require release notes?
- [ ] Yes
- [x] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidPlot/src/ApplicationWindow.cpp [hunk 1]
```diff
@@ -16762,8 +16775,19 @@ void ApplicationWindow::checkForProjectRecovery() {
...
+  try {
+    m_projectRecovery.attemptRecovery();
+  } catch (std::exception &e) {
+    std::string err{
+        "Project Recovery failed to recover this checkpoint. Details: "};
+    err.append(e.what());
+    g_log.error(err);
+    QMessageBox::information(this, "Could Not Recover",
+                             "We could not fully recover your work.\nMantid "
+                             "will continue to run normally now.",
```
Comment (Mantid-Matthew) [inline MantidPlot/src/ApplicationWindow.cpp]: Can we make clearer here that as we could not fully recover we are in fact doing no recovery and juts starting from a clean Mantid.

**Review by Matthew-Andrew — CHANGES_REQUESTED** [review]
Works for me as is a big improvement over the previous unhanded exception. Could we make the message displayed clearer as to exactly what state Mantid has been left in? i.e. I assume that no recovery is done at all ?

**Comment by DavidFair** [comment]
What would you recommend - I agree the text isn't brilliant but was unsure how to change it

**Comment by Mantid-Matthew** [comment]
There actually seems to be a more serious issue with this where certain errors will bypass the catch block and get to the last chance exception handler anyway. 
As an example
* Load a workspace
* Segfault
* go to your project recovery folder and corrupt the load line.
* Restart mantid

**Comment by Mantid-Matthew** [comment]
One more point is that currently if project recovery fails to complete I don't think new recording is ever started. Can we put the code to start recording a new history into the catch block so we return to a more sensible state.

**Comment by DavidFair** [comment]
I've put in a fix for the problem where a script status cannot be detected, so we continue to recover the GUI. 

There is a signal for failure emitted, however it is 6 layers deep in inheritance which means we would need to re-emit 6 times through the same mechanism.
I also considered using Poco::Observers however the additional code required to setup a notification centre was not ideal.

This fix will return a status when we synchronously execute to mark failure. In all other cases the flag on the scripting window is set to success (as we cannot tell asynchronously). Also to avoid us terminating from uncaught exceptions in the separate thread I simply log and return. 

The ideal solution going forwards would be to have the signal in `PythonScript` somehow connect to a slot in `ApplicationWindow` which triggers phase 2 of project recovery, or a failure reporting mechanism. This would also avoid us having to create a separate thread to run the code async from the GUI thread (which can block paint events), since we could rely on Qt informing us when we finish / fail the script.

**Review by Matthew-Andrew — APPROVED** [review]
This now blocks the last chance error handler in every case I tried. Also restarts the project recovery correctly after an error in the recovery scripts.

**Pull request merged**